eclipse-theia/theia
# Change default scope for `PreferenceService.set()`

#### What it does


Fixes #9152 by changing how the scope to which to apply a call to `PreferenceService.set()` is calculated. Formerly, if no scope was provided, `PreferenceScope.Folder` was used if a `URI` was supplied, `PreferenceScope.Workspace` otherwise. Now `PreferenceScope.User` is used instead of `PreferenceScope.Workspace`.

In addition, an error was previously thrown if `set` was called with a `sectionName` (`tasks`, `launch`) in User scope, but that is no longer appropriate because some section configurations are available in user scope, so now that check is delegated to `PreferenceSchema.isValidInScope()`. It could be argued that the check could be omitted, since an error will be thrown later in the pipeline (when preference changes are processed). 

#### How to test


1. 3. (You may want to set the environment variable THEIA_CONFIG_DIR to some temp dir)
2. Run the File > Auto Save menu item.
3. Observe that the setting is changed in user scope.
4. Create a command that calls e.g. `PreferenceService.set('tasks', {version: '2.0.0', tasks: []});`. E.g. add this to `PreferencesContribution.registerCommands()`:
```typescript
        commands.registerCommand({id: 'test-setting-preferences', label: 'Set a preference using PreferenceService.set'}, {
            execute: () => this.preferenceService.set('tasks', {version: '2.0.0', tasks: []}),
        });
```
5. Observe that it that setting takes effect in user scope and no error is logged.
6. Set some other preferences using the preference service and see how it goes.
#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>
